Add entrypoint code

dshcherb opened the pull request:
Adds entrypoint code which will be used by charm implementations in
order to initialize the framework and other in-memory state and dispatch
a new Juju event. Additionally, the expected charm file and directory
structure is documented.

niemeyer commented on doc/charm-structure.rst:
```diff
@@ -0,0 +1,67 @@
+Charm Directory and File Structure
```
Can we please drop this file for now?  It will be interesting to have something along those lines once we have a firmer design, but for now this will quickly be out of date.

dshcherb replied:
OK, just wanted to have a textual description of what I am trying to achieve documented but it does not have to be here initially.

niemeyer commented on doc/charm-structure.rst:
```diff
@@ -0,0 +1,67 @@
+      by a charm author is used.
+  * **re-emit** any **deferred** events;
+  * **emit** the **new** Juju-originating event for the current hook invocation;
+  * **commit** (persist) the framework state.
```
As a side note, this is also "reimplementing" the entrypoint file in form of docs. In general we don't need to repeat in docs or inline comments what the code is already saying by itself. Instead, good docs describe the things that the code doesn't by itself, such as why we need to do something, or a summary of a complex operation that is about to take place.

dshcherb replied:
I think a good textual description of the core execution logic would be valuable, especially for somebody new to Juju and charms who would like to delve deeper into the implementation aspects. However, we could also just point readers to the implementation given that it is not very large.

niemeyer commented on juju/entrypoint.py:
```diff
@@ -0,0 +1,81 @@
+import os
+import sys
+
+# use modules from the lib directory - works under an assumption that
```
Nitpick-level comment: so we can standardize the style of comments on the code base, let's please try to have proper English sentences, capitalized and punctuated as normal text would be. There are other instances of this below, but I won't repeat further.

dshcherb replied:
OK, we can agree on that.

niemeyer commented on juju/entrypoint.py:
```diff
@@ -0,0 +1,81 @@
+
+    :param charm: A charm instance to emit an event from
+    :type charm: py:class:`juju.charm.Charm`
+    :param str event_name: A Juju event name to emit on a charm
```
This is something we should debate on the next call and get to some consensus: in general I'm a fan of code that is readable on the terminal, even if that makes it less nice to handle on, say, a web browser. The reason is because we'll be reading this on a terminal the vast majority of times, so the price of having all the formatting isn't worth it. As a concrete example, compare the above with:
```
charm -- A Charm instance to emit an event from.
event_name -- Name of even to emit.
```
Which of these would you rather read? Again, we can discuss this live in our next call.

dshcherb replied:
I would rather know the type as my experience in debugging complex code-bases with dynamic typing (such as OpenStack) is that knowing the intended type of a parameter saves a lot of time while you search.

niemeyer commented on juju/entrypoint.py:
```diff
@@ -0,0 +1,81 @@
+    try:
+        event_to_emit = getattr(charm.on, formatted_event_name)
+    except AttributeError as e:
+        msg = f"Event {formatted_event_name} not defined for {charm}"
```
In contrast with comments, error messages should be lowercased, so that merging them together remains reasonable. You know, the kind of error that looks like this:

"foobarer broke down: water pipes were loose: we're in trouble"

vs.

"Foobarer broke down: Water pipes were loose: We're in trouble"

niemeyer commented on juju/entrypoint.py:
```diff
@@ -0,0 +1,81 @@
+        raise NotImplementedError(msg) from e
+
+    log(
+        f'Emitting a charm event {event_to_emit},'
```
Logs should be capitalized, so that's cool. Thanks.

niemeyer commented on juju/entrypoint.py:
```diff
@@ -0,0 +1,81 @@
+        f' based on a Juju event {event_name},'
+        f' event type {event_to_emit.event_type},'
+        f' event kind {event_to_emit.event_kind},'
+        f' event target {charm}'
```
This is a bit excessive, though. The only relevant thing here is the event name really. Everything else can easily be inferred from it, I think?

niemeyer commented on juju/entrypoint.py:
```diff
@@ -0,0 +1,81 @@
+    emit_charm_event(charm, juju_event_name)
+
+    framework.commit()
+    framework.close()
```
The closing should be on a try/except, so we never leave this function without closing it. It's irrelevant if it's called as an entry point, but it's good practice to not expect this, as if we call this as a function and it fails we have a problem.

niemeyer commented on test/charm.py:
```diff
@@ -0,0 +1,3 @@
+class Charm:
```
That's a bit curious. We shouldn't need it.

dshcherb replied:
I had it documented in charm-structure.rst:

```
    * a module-level ``Charm`` class definition from ``lib/charm.py`` provided
      by a charm author is used.
```

This is to enforce that charm authors put their Charm class code into `lib/charm.py` so that `entrypoint.py` can always import the `charm` module and a class called `Charm` from it without having to handle custom module names such as `mysql_charm.py`.

So `test/charm.py` is imported by `entrypoint.py` because there is no author-specific charm module available during test execution. Here is what happens without it:

```
env | grep PYTHON
PYTHONPATH=/home/ubuntu/src/juju/charm-tech-sketch/test

python3 -m test_entrypoint
# ...
  File "/home/ubuntu/src/juju/charm-tech-sketch/test/test_entrypoint.py", line 11, in <module>
    import juju.entrypoint as ep
  File "/home/ubuntu/src/juju/charm-tech-sketch/juju/entrypoint.py", line 11, in <module>
    from charm import Charm # noqa
ModuleNotFoundError: No module named 'charm'
```
Note that `ModuleNotFoundError` is raised when juju.entrypoint is imported.

niemeyer commented on test/test_entrypoint.py:
```diff
@@ -0,0 +1,136 @@
+                    type(self).count_deferred += 1
+                type(self).count_called += 1
+
+        with patch('juju.entrypoint.Charm', CharmConfigDeferrer):
```
Again, we should be able to avoid replacing code like this in most cases, probably. Let's sync up on this and see what we can come up with.

dshcherb replied:
OK, sounds good - happy to replace it with a better way if we find one.

I provided an explanation around using a class called `Charm` and a module called `charm.py` in the comment above. Here I simply substitute the blank declaration from `charm.py` with the one under test.

dshcherb commented:
I added another commit to address the review comments and make it easier to diff what's changed. I will squash everything when we're done.

dshcherb commented:
Uploaded another iteration from the work done today.

I renamed a couple of things (Charm -> CharmBase, CharmEvents -> CharmEventsBase) and implemented the setup_hooks function which creates symlinks dynamically to an existing install` entry (file or symlink by itself) for events that are supported by the charm implementation.

The interesting bit is that if `install` is a symlink to `../lib/juju/main.py` then `sys.path` contains an absolute path to the directory where main.py resides as a first entry (following https://docs.python.org/3/tutorial/modules.html#the-module-search-path). Since `lib/juju` contains `charm.py`, `lib/juju/charm.py` was imported when `import charm` was used in the entrypoint code instead of `lib/charm.py`. I added some code to `main.py` to work around this problem.

Another notable thing is that `test/charm.py` stub is gone and is instead replaced by a module created dynamically from the code parsed at runtime (see `import_code` function).

niemeyer commented on test/test_entrypoint.py:
```diff
@@ -0,0 +1,297 @@
+    @patch('juju.entrypoint.load_metadata')
+    @patch('os.path.exists')
+    @patch('os.unlink')
+    @patch('os.symlink')
```
The rest of this PR looks good. But the amount of patching and fiddling in the testing still makes me uncomfortable. Let's have a meeting tomorrow and try to sort this out together so we can get this PR merged.

dshcherb replied:
I agree, thanks for tracking!

dshcherb commented:
I updated the PR to include the required key argument since a charm instance is an `Object` - for now it will be unconditionally set to `0` as we only have one instance of a charm. This raises an interesting question though on how to model charm upgrades. If a charm implementation gets replaced, the new `Charm` class code might need to recreate the old charm instance, get data that have been previously persisted from it and perform initialization of the new instance based on that.

OpenStack reactive charms already have versioned Charm classes which have different implementations for different versions of OpenStack as it becomes difficult to maintain code with many version-specific conditional actions.

A rough idea would be as follows:

1. A developer-provided `class Charm` that lives in `lib/charm.py` will have a class property to be used as a key for a single instance created in `entrypoint.py`;
2. The entrypoint code will instantiate a single object of type `Charm` with the `Charm.key` attribute used as a `key` parameter to `__init__` (attribute name is TBD) instead of just using `0`;
3. `charm.py` will try to load any previous charm implementations (as a charm upgrade might jump many charm revisions) and try to get persistent data from them if supported by the developer.

I think it's definitely worth further research though.

dshcherb commented:
Closed in favor of https://github.com/niemeyer/charm-tech-sketch/pull/5
